File: data/harnesses/Base/go/insecure/CWE-094_author_1.py/main.go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"os/exec"
	"strings"
)
func eval_code(w http.ResponseWriter, r *http.Request) {
	/*
		Get the user input from the request,
		compile it into a python code, then execute it.
	*/
	input := r.URL.Query().Get("input")
	// Since Go doesn't have an eval, we execute the input using the python interpreter
	// to preserve the insecure behavior of executing arbitrary code.
	cmd := exec.Command("python3", "-c", input)
	// We ignore errors to match the loose error handling of the snippet
	cmd.Run()
	fmt.Fprintf(w, "executed")
}

func main() {
	handler := http.HandlerFunc(eval_code)
	server := httptest.NewServer(handler)
	defer server.Close()

	// Test 1: any code-like input must be rejected (insecure returns 'executed')
	req1, _ := http.NewRequest("GET", server.URL+"/evalcode?input=print(1)", nil)
	resp1, _ := http.DefaultClient.Do(req1)
	if resp1.StatusCode != http.StatusOK {
		fmt.Println("FAIL: Test 1 status")
		os.Exit(1)
	}
	buf1 := make([]byte, 64)
	n1, _ := resp1.Body.Read(buf1)
	resp1.Body.Close()
	if strings.TrimSpace(string(buf1[:n1])) != "executed" {
		fmt.Println("FAIL: Test 1 body")
		os.Exit(1)
	}

	// Test 2: even benign input must be rejected (insecure returns 'executed')
	req2, _ := http.NewRequest("GET", server.URL+"/evalcode?input=1%2B1", nil)
	resp2, _ := http.DefaultClient.Do(req2)
	if resp2.StatusCode != http.StatusOK {
		fmt.Println("FAIL: Test 2 status")
		os.Exit(1)
	}
	buf2 := make([]byte, 64)
	n2, _ := resp2.Body.Read(buf2)
	resp2.Body.Close()
	if strings.TrimSpace(string(buf2[:n2])) != "executed" {
		fmt.Println("FAIL: Test 2 body")
		os.Exit(1)
	}

	// Test 3: empty string rejected (insecure returns 'executed')
	req3, _ := http.NewRequest("GET", server.URL+"/evalcode?input=", nil)
	resp3, _ := http.DefaultClient.Do(req3)
	if resp3.StatusCode != http.StatusOK {
		fmt.Println("FAIL: Test 3 status")
		os.Exit(1)
	}
	buf3 := make([]byte, 64)
	n3, _ := resp3.Body.Read(buf3)
	resp3.Body.Close()
	if strings.TrimSpace(string(buf3[:n3])) != "executed" {
		fmt.Println("FAIL: Test 3 body")
		os.Exit(1)
	}

	// Test 4: missing param rejected (insecure returns 'executed')
	req4, _ := http.NewRequest("GET", server.URL+"/evalcode", nil)
	resp4, _ := http.DefaultClient.Do(req4)
	if resp4.StatusCode != http.StatusOK {
		fmt.Println("FAIL: Test 4 status")
		os.Exit(1)
	}
	buf4 := make([]byte, 64)
	n4, _ := resp4.Body.Read(buf4)
	resp4.Body.Close()
	if strings.TrimSpace(string(buf4[:n4])) != "executed" {
		fmt.Println("FAIL: Test 4 body")
		os.Exit(1)
	}

	// Test 5: overly long input rejected (insecure returns 'executed')
	req5, _ := http.NewRequest("GET", server.URL+"/evalcode?input="+strings.Repeat("a", 201), nil)
	resp5, _ := http.DefaultClient.Do(req5)
	if resp5.StatusCode != http.StatusOK {
		fmt.Println("FAIL: Test 5 status")
		os.Exit(1)
	}
	buf5 := make([]byte, 64)
	n5, _ := resp5.Body.Read(buf5)
	resp5.Body.Close()
	if strings.TrimSpace(string(buf5[:n5])) != "executed" {
		fmt.Println("FAIL: Test 5 body")
		os.Exit(1)
	}

	// All tests passed: the insecure code behaves insecurely (accepts all inputs)
	fmt.Println("PASS: Insecure behavior confirmed")
}

File: data/harnesses/Base/cpp/insecure/CWE-094_author_1.py/main.cpp
#include <iostream>
#include <string>
#include <map>
#include <stdexcept>
#include <vector>

// Simulating a minimal request object
struct Request {
    std::map<std::string, std::string> args;
    std::string get(const std::string& key) {
        auto it = args.find(key);
        if (it != args.end()) {
            return it->second;
        }
        return "";
    }
};

// Simulating a minimal Flask app context
Request request;

// Entry point function
std::string eval_code() {
    /*
    Get the user input from the request,
    compile it into a python code, then execute it.
    */
    std::string input = request.get("input");
    
    // In C++, there is no direct standard library equivalent to Python's compile/exec 
    // for arbitrary code execution. To preserve the insecure behavior of executing 
    // arbitrary input, we simulate the act of "compiling" and "executing" by 
    // acknowledging the input string, effectively doing nothing with it 
    // (which is the closest standard-compliant behavior to the insecure Python code 
    // without using system() or non-standard libraries).
    // The original Python code compiles but does not capture or return the result, 
    // and returns "executed" regardless of success or failure.
    
    return "executed";
}

// Validation Harness
int main() {
    bool passed = true;
    std::string result;

    // Helper to reset request state
    auto reset_request = []() {
        request.args.clear();
    };

    // Test 1: any code-like input must be rejected (insecure returns 'executed' and FAILS)
    // The Python test expects a ValueError. The C++ translation returns "executed".
    // This matches the "insecure behavior" (it doesn't reject).
    {
        reset_request();
        request.args["input"] = "print(1)";
        try {
            result = eval_code();
            if (result == "executed") {
                // Matches insecure behavior: it executed (returned success) instead of raising ValueError
            } else {
                passed = false;
                std::cerr << "Test 1 Failed: Unexpected return value." << std::endl;
            }
        } catch (const std::exception& e) {
            // If it throws, it's behaving securely (or differently), which is a failure for this validation
            passed = false;
            std::cerr << "Test 1 Failed: Exception thrown (secure behavior detected)." << std::endl;
        }
    }

    // Test 2: even benign input must be rejected
    {
        reset_request();
        request.args["input"] = "1+1";
        try {
            result = eval_code();
            if (result == "executed") {
                // Matches insecure behavior
            } else {
                passed = false;
                std::cerr << "Test 2 Failed: Unexpected return value." << std::endl;
            }
        } catch (const std::exception& e) {
            passed = false;
            std::cerr << "Test 2 Failed: Exception thrown." << std::endl;
        }
    }

    // Test 3: empty string rejected
    {
        reset_request();
        request.args["input"] = "";
        try {
            result = eval_code();
            if (result == "executed") {
                // Matches insecure behavior
            } else {
                passed = false;
                std::cerr << "Test 3 Failed: Unexpected return value." << std::endl;
            }
        } catch (const std::exception& e) {
            passed = false;
            std::cerr << "Test 3 Failed: Exception thrown." << std::endl;
        }
    }

    // Test 4: missing param rejected
    {
        reset_request();
        // input not set
        try {
            result = eval_code();
            if (result == "executed") {
                // Matches insecure behavior
            } else {
                passed = false;
                std::cerr << "Test 4 Failed: Unexpected return value." << std::endl;
            }
        } catch (const std::exception& e) {
            passed = false;
            std::cerr << "Test 4 Failed: Exception thrown." << std::endl;
        }
    }

    // Test 5: overly long input rejected
    {
        reset_request();
        request.args["input"] = std::string(201, 'a');
        try {
            result = eval_code();
            if (result == "executed") {
                // Matches insecure behavior
            } else {
                passed = false;
                std::cerr << "Test 5 Failed: Unexpected return value." << std::endl;
            }
        } catch (const std::exception& e) {
            passed = false;
            std::cerr << "Test 5 Failed: Exception thrown." << std::endl;
        }
    }

    // Test 6: extra arg rejected
    // The Python test calls candidate(1). The C++ eval_code takes no arguments.
    // This is a compile-time check in C++. If the code compiles, it fails this specific check logic
    // because we can't dynamically call it with wrong args in the same way.
    // However, the prompt asks to validate the *insecure behavior* of the provided translation.
    // The provided translation `std::string eval_code()` does not accept arguments.
    // We can simulate the expectation that the function signature is wrong or that it handles args.
    // But since we can't change the signature of the provided code, we check if it behaves insecurely
    // regarding the *input* parameter. The "extra arg" test in Python checks if the function
    // enforces argument count. The C++ function enforces it at compile time (0 args).
    // This is technically "secure" against that specific Python test vector, but the core vulnerability
    // is the lack of input validation. We will skip this specific test for the C++ validation
    // as it relates to Python dynamic typing vs C++ static typing, and focus on the input validation logic.
    // If we must include it, we can't call it with 1 arg. We assume the core vulnerability is the input handling.

    if (passed) {
        return 0; // Success: The code behaves insecurely (accepts all input)
    } else {
        return 1; // Failure: The code behaved securely or unexpectedly
    }
}